node/bft/src/primary.rs

```diff
@@ -94,12 +94,12 @@ pub struct Primary<N: Network> {
     proposed_batch: Arc<ProposedBatch<N>>,
     /// The timestamp of the most recent proposed batch.
     latest_proposed_batch_timestamp: Arc<RwLock<i64>>,
-    /// The recently-signed batch proposals (a map from the address to the round, timestamp, batch ID, and signature).
-    signed_proposals: Arc<RwLock<HashMap<Address<N>, (u64, i64, Field<N>, Signature<N>)>>>,
+    /// The recently-signed batch proposals.
+    signed_proposals: Arc<RwLock<SignedProposals<N>>>,
     /// The spawned handles.
     handles: Arc<Mutex<Vec<JoinHandle<()>>>>,
     /// The lock for propose_batch.
-    propose_lock: Arc<TMutex<(u64, i64)>>,
+    propose_lock: Arc<TMutex<u64>>,
```

Comment:
```suggestion
    propose_lock: Arc<AtomicU64>,
```

Reply:
Though I now see that it also serves as a locking mechanism... `clippy` might complain about it, though.